zerver/views/read_receipts.py

```diff
@@ -0,0 +1,33 @@
+from django.http.request import HttpRequest
+from django.http.response import HttpResponse
+
+from zerver.lib.message import access_message
+from zerver.lib.request import REQ, has_request_variables
+from zerver.lib.response import json_success
+from zerver.lib.validator import to_non_negative_int
+from zerver.models import UserMessage, UserProfile
+
+
+@has_request_variables
+def read_receipts(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    message_id: int = REQ(converter=to_non_negative_int, path_only=True),
+) -> HttpResponse:
+
+    # Verify user has access to message.
+    access_message(user_profile, message_id)
+
+    user_ids = (
+        UserMessage.objects.filter(
+            message_id=message_id,
+            user_profile__is_active=True,
+            user_profile__send_read_receipts=True,
+        )
+        .extra(
+            where=[UserMessage.where_read()],
```

Comment:
We likely want to exclude any rows with `UserMessage.flags.historical` set, since those will be cases where the user with said flag was not a recipient of the message at the time it was sent.

Reply:
It might look like a bug when a user has reacted to a message thus creating a `UserMessage` with `historical` flag but do not appear in read receipts.

For now, I didn't commit any changes in this respect thinking it might need some discussion. If we did want to exclude these, is it correct to use: `.exclude(flags=UserMessage.flags.historical)`?

Reply:
Yes, we should make this change. I think that's how `django-bitfield` works, but worth checking its documentation.

Reply:
Also, please extend one of the tests to cover this case, as well as the excluding the message sender in the API response.

(Possibly just extending an existing test with more setup).

Reply:
Yeah, added that change. I came up with that `.exclude(flags=UserMessage.flags.historical)` looking at `django-bitfield` docs but was confused by us using `where_unread` etc and was wondering if we should be doing something like that.

For tests, added `test_historical_usermessages_read_flag_isnt_considered` taking advantage of reactions to create a historical `UserMessage`. Could you check if it looks right?

Excluding message sender case was already covered in the last lines of `test_stream_message` and `test_personal_message`.

Reply:
I think this is probably moot, because we seem to have decided to not filter historical anyway :)